spec/index.bs

```diff
@@ -3545,9 +3666,85 @@ interface mixin GPUProgrammablePassEncoder {
 };
 </script>
 
+{{GPUProgrammablePassEncoder}} has the following internal slots:
+
+<dl dfn-type=attribute dfn-for="GPUProgrammablePassEncoder">
+    : <dfn>\[[debug_group_stack]]</dfn> of type `sequence<DOMString>`.
+    ::
+        A stack of active debug group labels.
+</dl>
+
+## Debug Markers ## {#programmable-passes-debug-markers}
+
+Similar to Command encoders, programmable pass encoders provide methods to apply debug labels to
+groups of commands or insert a single label into the command sequence. Debug groups can be nested to
+create a hierarchy of labeled commands. These labels MAY be passed to the native API backends for
+tooling, MAY be used by the user agent's internal tooling, or MAY be a no-op when such tooling is
+not available or applicable.
+
 Debug groups in a {{GPUCommandEncoder}} or {{GPUProgrammablePassEncoder}}
 must be well nested.
 
+### <dfn method for=GPUProgrammablePassEncoder>pushDebugGroup(groupLabel)</dfn> ### {#GPUProgrammablePassEncoder-pushDebugGroup}
+
+<div algorithm="GPUProgrammablePassEncoder.pushDebugGroup">
+    <strong>|this|:</strong> of type {{GPUProgrammablePassEncoder}}.
+
+    **Arguments:**
+      - {{DOMString}} |groupLabel|
+
+    **Returns:** void
+
+    Marks the beginning of a labeled group of commands for the {{GPUProgrammablePassEncoder}}.
+
+    |groupLabel| defines the label for the command group.
+
+    On the [=Device timeline=], the following steps occur:
+
+        - If the user agent can fulfill the request:
+
+            - push |groupLabel| onto then end of |this|.{{GPUProgrammablePassEncoder/[[debug_group_stack]]}}.
+</div>
+
+### <dfn method for=GPUProgrammablePassEncoder>popDebugGroup()</dfn> ### {#GPUProgrammablePassEncoder-popDebugGroup}
+
+<div algorithm="GPUProgrammablePassEncoder.popDebugGroup">
+    <strong>|this|:</strong> of type {{GPUProgrammablePassEncoder}}.
+
+    **Returns:** void
+
+    Marks the end of a labeled group of commands for the {{GPUProgrammablePassEncoder}}.
+
+    On the [=Device timeline=], the following steps occur:
+
+        - If the user agent can fulfill the request and
+            the [$GPUProgrammablePassEncoder.popDebugGroup/Valid Usage$] rules are met:
+
+            - pop an entry off the end of |this|.{{GPUProgrammablePassEncoder/[[debug_group_stack]]}}.
+
+    <div class=validusage dfn-for=GPUProgrammablePassEncoder.popDebugGroup>
+        <dfn abstract-op>Valid Usage</dfn>
+
+        For |this|:
+        - |this|.{{GPUProgrammablePassEncoder/[[debug_group_stack]]}}.length must be greater than 0.
+    </div>
+
+</div>
+
+### <dfn method for=GPUProgrammablePassEncoder>insertDebugMarker(markerLabel)</dfn> ### {#GPUProgrammablePassEncoder-insertDebugMarker}
+
+<div algorithm="GPUProgrammablePassEncoder.insertDebugMarker">
+
+    **Arguments:**
+        - {{DOMString}} |markerLabel|
+
+    **Returns:** void
+
+    Inserts a single debug marker label into the {{GPUProgrammablePassEncoder}}'s commands sequence .
```

Comment:
what are the validation differences with the markers in `GPUCommandEncoder`?

Reply:
The only difference I've found is that the command encoder has to check to see if a pass is currently being recorded, while the passes do not for obvious reasons. :) Also the [[debug_group_stack]] link would need to be disambiguated. That applies to all the debug commands, actually, and is really the only thing keeping them from being defined in a single place.

I could write the validation logic to be something along the lines of:

```
For |this|:
    If |this| is a {{GPUCommandEncoder}}:
        - |this|.{{GPUCommandEncoder/[[state]]}} must be {{encoder state/open}}.
        - |this|.{{GPUCommandEncoder/[[debug_group_stack]]}}.length must be greater than 0.
    Otherwise if |this| is a {{GPUProgrammablePassEncoder}}:
        - |this|.{{GPUProgrammablePassEncoder/[[debug_group_stack]]}}.length must be greater than 0.
```

That felt more awkward to me, but I'm happy to write it either way.

Reply:
What about moving the logic into some base `GPUEncoderBase` interface? We shouldn't need this switch then.

Reply:
The logic would still be slightly different due to the encoder state check. One way to address that would be to have the encode state reside on the `GPUEncoderBase` and indicate that the pass encoders can only be `open` or `closed`.

Happy to go that route if the group wants. Assuming the answer is "yes" would you want that to be part of this PR or a follow up?

Reply:
That suggestion sounds pretty good to me, but on the CommandEncoder we need to also make sure there's no pass active. So I think CommandEncoder.state needs to be `open|render|compute|closed`, while pass encoder states need to be `open|closed`.

Probably do a followup since it may get a bit fidgety.